addons/pos_loyalty/models/pos_session.py

```diff
@@ -9,60 +9,38 @@
 class PosSession(models.Model):
     _inherit = 'pos.session'
 
-    def _pos_ui_models_to_load(self):
-        result = super()._pos_ui_models_to_load()
-        if self.config_id._get_program_ids():
-            result += [
-                'loyalty.program',
-                'loyalty.rule',
-                'loyalty.reward',
-            ]
-        return result
-
-    def _loader_params_loyalty_program(self):
-        return {
-            'search_params': {
+    def _load_data_params(self, config_id):
+        params = super()._load_data_params(config_id)
+        params['product.product']['fields'].append('all_product_tag_ids')
+        params.update({
+            'loyalty.program': {
                 'domain': [('id', 'in', self.config_id._get_program_ids().ids)],
                 'fields': [
                     'name', 'trigger', 'applies_on', 'program_type', 'pricelist_ids', 'date_from',
                     'date_to', 'limit_usage', 'max_usage', 'is_nominative', 'portal_visible',
-                    'portal_point_name', 'trigger_product_ids',
+                    'portal_point_name', 'trigger_product_ids', 'rule_ids', 'reward_ids'
                 ],
             },
-        }
-
-    def _loader_params_loyalty_rule(self):
-        return {
-            'search_params': {
+            'loyalty.rule': {
                 'domain': [('program_id', 'in', self.config_id._get_program_ids().ids)],
                 'fields': ['program_id', 'valid_product_ids', 'any_product', 'currency_id',
                     'reward_point_amount', 'reward_point_split', 'reward_point_mode',
                     'minimum_qty', 'minimum_amount', 'minimum_amount_tax_mode', 'mode', 'code'],
-            }
-        }
-
-    def _loader_params_loyalty_reward(self):
-        return {
-            'search_params': {
+            },
+            'loyalty.reward': {
                 'domain': [('program_id', 'in', self.config_id._get_program_ids().ids)],
                 'fields': ['description', 'program_id', 'reward_type', 'required_points', 'clear_wallet', 'currency_id',
                     'discount', 'discount_mode', 'discount_applicability', 'all_discount_product_ids', 'is_global_discount',
                     'discount_max_amount', 'discount_line_product_id',
                     'multi_product', 'reward_product_ids', 'reward_product_qty', 'reward_product_uom_id', 'reward_product_domain'],
-            }
-        }
-
-    def _get_pos_ui_loyalty_program(self, params):
-        return self.env['loyalty.program'].search_read(**params['search_params'])
-
-    def _get_pos_ui_loyalty_rule(self, params):
-        return self.env['loyalty.rule'].search_read(**params['search_params'])
+            },
+            'loyalty.card': {
+                'domain': [],
+                'fields': ['partner_id', 'code', 'points', 'program_id'],
+            },
+        })
```

Comment:
If we load loyalty.card like this, we'll be loading it twice because it's also done in '_set_loyalty_cards'. Also, we should only load the card that are coming from 'loyalty' programs.